core/store/src/trie/split_state.rs

```diff
@@ -0,0 +1,372 @@
+    ///
+    /// # Errors
+    /// StorageError if the storage is corrupted
+    pub fn get_trie_items_for_part(
+        &self,
+        part_id: u64,
+        num_parts: u64,
+        state_root: &StateRoot,
+    ) -> Result<Vec<TrieItem>, StorageError> {
+        assert!(part_id < num_parts);
+        assert!(self.storage.as_caching_storage().is_some());
+
+        let path_begin = self.find_path_for_part_boundary(state_root, part_id, num_parts)?;
+        let path_end = self.find_path_for_part_boundary(state_root, part_id + 1, num_parts)?;
+        self.iter(state_root)?.get_trie_items(&path_begin, &path_end)
+    }
+}
+
+impl ShardTries {
+    /// Apply `changes` to build states for new shards
+    /// `state_roots` contains state roots for the new shards
+    /// Ignore changes on DelayedReceipts or DelayedReceiptsIndices
+    /// Update `store_update` and return new state_roots
+    /// used for building states for new shards in resharding
+    pub fn apply_changes_to_new_states<'a>(
+        &self,
+        state_roots: HashMap<ShardUId, StateRoot>,
+        changes: Vec<(Vec<u8>, Option<Vec<u8>>)>,
+        key_to_shard_id: Box<dyn Fn(&Vec<u8>) -> Option<ShardUId> + 'a>,
+    ) -> Result<(StoreUpdate, HashMap<ShardUId, StateRoot>), StorageError> {
+        let mut changes_by_shard = HashMap::new();
+        for (raw_key, value) in changes.into_iter() {
+            if let Some(new_shard_uid) = key_to_shard_id(&raw_key) {
+                assert!(
+                    state_roots.contains_key(&new_shard_uid),
+                    "Cannot find state roots for shard {:?} for key {}. Only has state roots for shards {:?}",
+                    new_shard_uid,
+                    from_utf8(&raw_key).unwrap(),
+                    state_roots.keys(),
+                );
+                changes_by_shard
+                    .entry(new_shard_uid)
+                    .or_insert_with(Vec::new)
+                    .push((raw_key, value));
+            }
+        }
+        let mut new_state_roots = state_roots.clone();
+        let mut store_update = StoreUpdate::new_with_tries(self.clone());
+        for (shard_uid, changes) in changes_by_shard.into_iter() {
+            let trie = self.get_trie_for_shard(shard_uid.clone());
+            let trie_changes = trie.update(&state_roots[&shard_uid], changes.into_iter())?;
+            let (update, state_root) = self.apply_all(&trie_changes, shard_uid.clone())?;
+            new_state_roots.insert(shard_uid, state_root);
+            store_update.merge(update);
+        }
+        Ok((store_update, new_state_roots))
+    }
+
+    pub fn apply_delayed_receipts_to_split_states<'a>(
+        &self,
+        state_roots: &HashMap<ShardUId, StateRoot>,
+        receipts: &Vec<Receipt>,
+        account_id_to_shard_id: Box<dyn Fn(&AccountId) -> ShardUId + 'a>,
+    ) -> Result<(StoreUpdate, HashMap<ShardUId, StateRoot>), StorageError> {
+        let mut trie_updates: HashMap<_, _> = HashMap::new();
```

Comment:
```suggestion
        let mut trie_updates = HashMap::new();
```

the type signature doesn't add any extra info